File: Application/src/main/java/com/shareconnect/SettingsActivity.java
package com.shareconnect;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.MenuItem;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

public class SettingsActivity extends AppCompatActivity {
    private static final int THEME_SELECTION_REQUEST = 1001;
    private ThemeManager themeManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        // Apply current theme before calling super.onCreate()
        themeManager = ThemeManager.getInstance(this);
        themeManager.applyTheme(this);
        
        super.onCreate(savedInstanceState);
        
        setContentView(R.layout.settings_activity);
        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.settings, new SettingsFragment())
                .commit();
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onResume() {
        super.onResume();
        // Check if theme has changed and recreate activity if needed
        if (themeManager != null && themeManager.hasThemeChanged()) {
            themeManager.resetThemeChangedFlag();
            recreate();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        // If theme was changed, recreate this activity to apply the new theme
        if (requestCode == THEME_SELECTION_REQUEST && resultCode == RESULT_OK) {
            recreate();
        }
    }

    public static class SettingsFragment extends PreferenceFragmentCompat {
        @Override
        public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
            setPreferencesFromResource(R.xml.root_preferences, rootKey);
            
            Preference profilesPreference = findPreference("server_profiles");
            if (profilesPreference != null) {
                profilesPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                    @Override
                    public boolean onPreferenceClick(Preference preference) {
                        Intent intent = new Intent(getContext(), ProfilesActivity.class);
                        startActivity(intent);
                        return true;
                    }
                });
            }
            
            Preference themePreference = findPreference("theme_selection");
            if (themePreference != null) {
                themePreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                    @Override
                    public boolean onPreferenceClick(Preference preference) {
                        Intent intent = new Intent(getContext(), ThemeSelectionActivity.class);
                        // Start for result to know when theme changes
                        startActivityForResult(intent, THEME_SELECTION_REQUEST);
                        return true;
                    }
                });
            }
        }
    }
}

File: Application/src/main/kotlin/com/shareconnect/SettingsActivity.kt
package com.shareconnect

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.redelf.commons.logging.Console

class SettingsActivity : AppCompatActivity() {
    private var themeManager: ThemeManager? = null
    private var isFirstRun = false

    override fun onCreate(savedInstanceState: Bundle?) {
        // Apply current theme before calling super.onCreate()
        themeManager = ThemeManager.getInstance(this)
        themeManager!!.applyTheme(this)

        super.onCreate(savedInstanceState)

        // Enable edge-to-edge display
        window.statusBarColor = android.graphics.Color.TRANSPARENT
        window.navigationBarColor = android.graphics.Color.TRANSPARENT

        // Check if this is the first run
        isFirstRun = intent.getBooleanExtra("first_run", false)

        setContentView(R.layout.settings_activity)
        supportFragmentManager
            .beginTransaction()
            .replace(R.id.settings, SettingsFragment())
            .commit()
        val actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        if (isFirstRun) {
            // If this is the first run, we need to check if profiles were created
            val profileManager = ProfileManager(this)
            if (profileManager.hasProfiles()) {
                // Profiles were created, start MainActivity
                val intent = Intent(this, MainActivity::class.java)
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP)
                startActivity(intent)
                finish()
            } else {
                // No profiles created, just finish and let the app close
                finishAffinity() // This will finish all activities and properly close the app
            }
        } else {
            // Normal back behavior
            super.onBackPressed()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        // If theme was changed, recreate this activity to apply the new theme
        if (requestCode == THEME_SELECTION_REQUEST && resultCode == RESULT_OK) {
            Console.debug("Theme change detected, recreating activity")
            // Simply recreate the current activity instead of starting a new one
            recreate()
        }
    }

    fun startThemeSelection() {
        val intent = Intent(this, ThemeSelectionActivity::class.java)
        startActivityForResult(intent, THEME_SELECTION_REQUEST)
    }

    class SettingsFragment : PreferenceFragmentCompat() {
        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
            setPreferencesFromResource(R.xml.root_preferences, rootKey)

            val profilesPreference = findPreference<Preference>("server_profiles")
            if (profilesPreference != null) {
                profilesPreference.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                    val intent = Intent(context, ProfilesActivity::class.java)
                    startActivity(intent)
                    true
                }
            }

            val themePreference = findPreference<Preference>("theme_selection")
            if (themePreference != null) {
                themePreference.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                    // Call the parent activity's method to start theme selection
                    (activity as? SettingsActivity)?.startThemeSelection()
                    true
                }
            }
        }
    }

    companion object {
        private const val THEME_SELECTION_REQUEST = 1001
    }
}